core/src/main/scala/io/snappydata/TableStatsProviderService.scala

```diff
@@ -0,0 +1,293 @@
+              val x = key.getKeyColumn(2).getInt
+              val currentBucketRegion = itr.getHostedBucketRegion
+              if (x == JDBCSourceAsStore.STATROW_COL_INDEX) {
+                val v = currentBucketRegion.get(key)
+                if (v ne null) {
+                  val currentVal = v.asInstanceOf[Array[Array[Byte]]]
+                  val rowFormatter = container.
+                      getRowFormatter(currentVal(0))
+                  val statBytes = rowFormatter.getLob(currentVal, 4)
+                  val unsafeRow = new UnsafeRow(numColumnsInStatBlob);
+                  unsafeRow.pointTo(statBytes, Platform.BYTE_ARRAY_OFFSET,
+                    statBytes.length);
+                  rowsInColumnBatch += unsafeRow.getInt(ColumnStatsSchema.COUNT_INDEX_IN_SCHEMA);
+                }
+              }
+            }
+            pr.getPrStats.setPRNumRowsInColumnBatches(rowsInColumnBatch)
+          }
+        }
+      }
+    }
+  }
+
+  def getAggregatedStatsOnDemand: (Map[String, SnappyRegionStats],
+      Map[String, SnappyIndexStats]) = {
+    val snc = this.snc
+    if (snc == null) return (Map.empty, Map.empty)
+    val (tableStats, indexStats) = getStatsFromAllServers
+
+    val aggregatedStats = scala.collection.mutable.Map[String, SnappyRegionStats]()
+    val aggregatedStatsIndex = scala.collection.mutable.Map[String, SnappyIndexStats]()
+    if (!doRun) return (Map.empty, Map.empty)
+    // val samples = getSampleTableList(snc)
+    tableStats.foreach { stat =>
+      aggregatedStats.get(stat.getRegionName) match {
+        case Some(oldRecord) =>
+          aggregatedStats.put(stat.getRegionName, oldRecord.getCombinedStats(stat))
+        case None =>
+          aggregatedStats.put(stat.getRegionName, stat)
+      }
+    }
+
+    indexStats.foreach { stat =>
+      aggregatedStatsIndex.put(stat.getIndexName, stat)
+    }
+    (Utils.immutableMap(aggregatedStats), Utils.immutableMap(aggregatedStatsIndex))
+  }
+
+  /*
+  private def getSampleTableList(snc: SnappyContext): Seq[String] = {
+    try {
+      snc.sessionState.catalog
+          .getDataSourceTables(Seq(ExternalTableType.Sample)).map(_.toString())
+    } catch {
+      case tnfe: org.apache.spark.sql.TableNotFoundException =>
+        Seq.empty[String]
+    }
+  }
+  */
+
+  protected def getStatsFromAllServers: (Seq[SnappyRegionStats], Seq[SnappyIndexStats]) = {
+    var result = new java.util.ArrayList[SnappyRegionStatsCollectorResult]().asScala
+    val dataServers = GfxdMessage.getAllDataStores
+    if( dataServers != null && dataServers.size() > 0 ){
+      result = FunctionService.onMembers(dataServers)
+          .withCollector(new GfxdListResultCollector())
+          .execute(SnappyRegionStatsCollectorFunction.ID).getResult().
+          asInstanceOf[java.util.ArrayList[SnappyRegionStatsCollectorResult]]
+          .asScala
+    }
+    (result.flatMap(_.getRegionStats.asScala), result.flatMap(_.getIndexStats.asScala))
+  }
```

Comment:
See earlier comments. Why are there so many method bodies in this trait which belong to embedded impl?